File: app/src/main/java/com/wa2c/android/medoly/plugin/action/lyricsscraper/SiteActivity.java
package com.wa2c.android.medoly.plugin.action.lyricsscraper;

import android.app.ActionBar;
import android.app.LoaderManager;
import android.content.Context;
import android.content.CursorLoader;
import android.content.Intent;
import android.content.Loader;
import android.content.SharedPreferences;
import android.database.Cursor;
import android.net.Uri;
import android.os.Bundle;
import android.app.Activity;
import android.preference.PreferenceManager;
import android.text.TextUtils;
import android.view.KeyEvent;
import android.view.Menu;
import android.view.MenuItem;
import android.view.MotionEvent;
import android.view.View;
import android.view.ViewGroup;
import android.widget.AbsListView;
import android.widget.AdapterView;
import android.widget.CursorAdapter;
import android.widget.ImageButton;
import android.widget.ListView;
import android.widget.RadioButton;
import android.widget.TextView;



public class SiteActivity extends Activity {

    /** Cursor loader ID. */
    private static final int CURSOR_LOADER_ID = 1;
    /** Cursor type. */
    private static final String CURSOR_TYPE = "cursor_type";
    /** Cursor type: site. */
    private static final int CURSOR_TYPE_SITE = 0;
    /** Cursor type: group. */
    private static final int CURSOR_TYPE_GROUP = 1;
    /** Group ID. */
    private static final String GROUP_ID = "group_id";

    /** Preferences. */
    private SharedPreferences preference;
    /** Cursor Adapter. */
    private SheetCursorAdapter cursorAdapter;
    /** Loader manager. */
    private LoaderManager loaderManager;


    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        this.preference = PreferenceManager.getDefaultSharedPreferences(this);
        setContentView(R.layout.activity_group);

        // action bar
        ActionBar actionBar = getActionBar();
        if (actionBar != null) {
            actionBar.setDisplayShowHomeEnabled(true);
            actionBar.setDisplayHomeAsUpEnabled(true);
        }

        // create loader manager
        Bundle bundle = new Bundle();
        bundle.putInt(CURSOR_TYPE, CURSOR_TYPE_GROUP);

        // create view
        cursorAdapter = new SheetCursorAdapter(this, CURSOR_TYPE_GROUP);
        ListView listView = (ListView) findViewById(R.id.groupListView);
        listView.setOnItemClickListener(new AdapterView.OnItemClickListener() {
            @Override
            public void onItemClick(AdapterView<?> parent, View view, int position, long id) {
                SheetCursorAdapter.ListViewHolder holder = (SheetCursorAdapter.ListViewHolder) view.getTag();
                if (cursorAdapter.getCursorType() == CURSOR_TYPE_GROUP) {
                    openSiteList(holder.GroupId);
                } else if (cursorAdapter.getCursorType() == CURSOR_TYPE_SITE) {
                    preference.edit().putString(getString(R.string.prefkey_selected_site_id), holder.SiteId).apply();
                    cursorAdapter.notifyDataSetChanged();
                }
            }
        });
        listView.setChoiceMode(AbsListView.CHOICE_MODE_SINGLE);
        listView.setAdapter(cursorAdapter);

        // create load manager
        loaderManager = this.getLoaderManager();
        loaderManager.restartLoader(CURSOR_LOADER_ID, bundle, cursorLoaderCallbacks);
    }



    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        getMenuInflater().inflate(R.menu.site_list, menu);
        return true;
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        switch (item.getItemId()) {
            case android.R.id.home:
                // home
                if (cursorAdapter.getCursorType() == CURSOR_TYPE_SITE) {
                    openGroupList();
                } else {
                    finish();
                }
                return true;
            case R.id.menu_update_list:
                // update list
                SpreadSheetReadTask task = new SpreadSheetReadTask(getApplicationContext());
                task.setOnPropertyActionListener(new SpreadSheetReadTask.SiteUpdateListener() {
                    @Override
                    public void onListUpdated(boolean isSucceeded) {
                        if (isSucceeded) {
                            openGroupList();
                            AppUtils.showToast(getApplicationContext(), R.string.message_renew_list_succeeded);
                        } else {
                            AppUtils.showToast(getApplicationContext(), R.string.message_renew_list_failed);
                        }
                    }
                });
                task.execute();
                return true;
            case R.id.menu_open_sheet:
                String sheetUrl = getString(R.string.sheet_uri, getString(R.string.sheet_id));
                Uri sheetUri = Uri.parse(sheetUrl);
                Intent browserIntent = new Intent(Intent.ACTION_VIEW, sheetUri);
                startActivity(browserIntent);
                return true;
        }
        return super.onOptionsItemSelected(item);
    }

    @Override
    public boolean dispatchKeyEvent(KeyEvent e) {
        switch (e.getKeyCode()) {
            // back key
            case KeyEvent.KEYCODE_BACK:
                if (cursorAdapter.getCursorType() == CURSOR_TYPE_SITE) {
                    openGroupList();
                    return true;
                }
                break;
        }
        return super.dispatchKeyEvent(e);
    }

    /**
     * Open site list.
     * @param groupId Group ID.
     */
    private void openSiteList(String groupId) {
        cursorAdapter.setCursorType(CURSOR_TYPE_SITE);
        Bundle bundle = new Bundle();
        bundle.putInt(CURSOR_TYPE, CURSOR_TYPE_SITE);
        bundle.putString(GROUP_ID, groupId);
        loaderManager.restartLoader(CURSOR_LOADER_ID, bundle, cursorLoaderCallbacks);
    }

    /**
     * Open group list.
     */
    private void openGroupList() {
        cursorAdapter.setCursorType(CURSOR_TYPE_GROUP);
        Bundle bundle = new Bundle();
        bundle.putInt(CURSOR_TYPE, CURSOR_TYPE_GROUP);
        loaderManager.restartLoader(CURSOR_LOADER_ID, bundle, cursorLoaderCallbacks);
    }


    /**
     * Cursor loader.
     */
    private final LoaderManager.LoaderCallbacks<Cursor> cursorLoaderCallbacks = new LoaderManager.LoaderCallbacks<Cursor>() {

        @Override
        public Loader<Cursor> onCreateLoader(int id, Bundle args) {
            if (args != null) {
                int type = args.getInt(CURSOR_TYPE, CURSOR_TYPE_GROUP);
                if (type == CURSOR_TYPE_GROUP) {
                    return new CursorLoader(
                            SiteActivity.this,
                            SiteProvider.GROUP_URI,
                            null,
                            null,
                            null,
                            GroupColumn.NAME.getColumnKey());
                } else if (type == CURSOR_TYPE_SITE) {
                    String groupId = args.getString(GROUP_ID, "-1");
                    return new CursorLoader(
                            SiteActivity.this,
                            SiteProvider.SITE_URI,
                            null,
                            SiteColumn.GROUP_ID.getColumnKey() + "=?",
                            new String[] { groupId },
                            SiteColumn.SITE_NAME.getColumnKey());
                }
            }
            return null;
        }

        @Override
        public void onLoadFinished(Loader<Cursor> loader, Cursor data) {
            if (loader.isReset() || loader.isAbandoned()) return;
            cursorAdapter.swapCursor(data);
            cursorAdapter.notifyDataSetChanged();
        }

        @Override
        public void onLoaderReset(Loader<Cursor> loader) {
            cursorAdapter.swapCursor(null);
        }
    };


    /**
     * Spreadsheet data cursor adapter.
     */
    private static class SheetCursorAdapter extends CursorAdapter {

        /** Cursor type. */
        private int cursorType;
        /** Set cursor type. */
        public void setCursorType(int cursorType) {
            this.cursorType = cursorType;
        }
        /** Get cursor type. */
        public int getCursorType() {
            return this.cursorType;
        }

        private SharedPreferences preference;

        public SheetCursorAdapter(Activity context, int cursorType) {
            super(context, null, true);
            this.cursorType = cursorType;
            this.preference = PreferenceManager.getDefaultSharedPreferences(context);
        }

        @Override
        public View newView(final Context context, final Cursor cursor, final ViewGroup parent) {
            final View view = View.inflate(context, R.layout.layout_site_item, null);
            final ListViewHolder holder = new ListViewHolder();
            holder.SelectRadioButton = (RadioButton)view.findViewById(R.id.siteSelectRadioButton);
            holder.TitleTextView = (TextView)view.findViewById(R.id.siteParamTitleTextView);
            holder.UriTextView = (TextView)view.findViewById(R.id.siteParamUriTextView);
            holder.LaunchImageButton = (ImageButton)view.findViewById(R.id.siteLaunchImageButton);
            view.setTag(holder);
            return view;
        }

        @Override
        public void bindView(final View view, final Context context, final Cursor cursor) {
            final ListViewHolder holder = (ListViewHolder) view.getTag();
            if (cursorType == CURSOR_TYPE_GROUP) {
                holder.SelectRadioButton.setVisibility(View.GONE);
                holder.LaunchImageButton.setVisibility(View.GONE);

                String title = cursor.getString(cursor.getColumnIndexOrThrow(GroupColumn.NAME.getColumnKey()));
                holder.GroupId = cursor.getString(cursor.getColumnIndexOrThrow(GroupColumn.GROUP_ID.getColumnKey()));
                holder.TitleTextView.setText(title);
                holder.UriTextView.setVisibility(View.GONE);

            } else {
                // ID
                holder.GroupId = cursor.getString(cursor.getColumnIndexOrThrow(SiteColumn.GROUP_ID.getColumnKey()));
                holder.SiteId = cursor.getString(cursor.getColumnIndexOrThrow(SiteColumn.SITE_ID.getColumnKey()));

                // select
                holder.SelectRadioButton.setVisibility(View.VISIBLE);
                if (holder.SiteId.equals(preference.getString(context.getString(R.string.prefkey_selected_site_id), "-1"))) {
                    holder.SelectRadioButton.setChecked(true);
                } else {
                    holder.SelectRadioButton.setChecked(false);
                }
                holder.SelectRadioButton.setOnTouchListener(new View.OnTouchListener() {
                    @Override
                    public boolean onTouch(View v, MotionEvent event) {
                        return view.onTouchEvent(event);
                    }
                });

                // title
                String title = cursor.getString(cursor.getColumnIndexOrThrow(SiteColumn.SITE_NAME.getColumnKey()));
                holder.TitleTextView.setText(title);
                holder.TitleTextView.setOnTouchListener(new View.OnTouchListener() {
                    @Override
                    public boolean onTouch(View v, MotionEvent event) {
                        return view.onTouchEvent(event);
                    }
                });

                // uri
                String uri = cursor.getString(cursor.getColumnIndexOrThrow(SiteColumn.SITE_URI.getColumnKey()));
                holder.UriTextView.setVisibility(View.VISIBLE);
                holder.UriTextView.setText(uri);
                holder.LaunchImageButton.setTag(uri);
                holder.LaunchImageButton.setVisibility(View.VISIBLE);
                if (!TextUtils.isEmpty(uri)) {
                    holder.LaunchImageButton.setEnabled(true);
                    holder.LaunchImageButton.setOnClickListener(new View.OnClickListener() {
                        @Override
                        public void onClick(View v) {
                            Uri sheetUri = Uri.parse((String) v.getTag());
                            Intent browserIntent = new Intent(Intent.ACTION_VIEW, sheetUri);
                            context.startActivity(browserIntent);
                        }
                    });
                } else {
                    holder.LaunchImageButton.setEnabled(false);
                }
            }
        }

        /**
         * ViewHolder for QueueListView.
         */
        private static class ListViewHolder {
            public String GroupId;
            public String SiteId;

            public RadioButton SelectRadioButton;
            public TextView TitleTextView;
            public TextView UriTextView;
            public ImageButton LaunchImageButton;
        }
    }

}

File: app/src/main/java/com/wa2c/android/medoly/plugin/action/lyricsscraper/activity/SiteActivity.kt
package com.wa2c.android.medoly.plugin.action.lyricsscraper.activity

import android.app.Activity
import android.content.Context
import android.content.Intent
import android.databinding.DataBindingUtil
import android.net.Uri
import android.os.Bundle
import android.view.*
import android.widget.AbsListView
import android.widget.AdapterView
import android.widget.ArrayAdapter
import com.wa2c.android.medoly.plugin.action.lyricsscraper.BuildConfig
import com.wa2c.android.medoly.plugin.action.lyricsscraper.R
import com.wa2c.android.medoly.plugin.action.lyricsscraper.databinding.ActivityGroupBinding
import com.wa2c.android.medoly.plugin.action.lyricsscraper.db.DbHelper
import com.wa2c.android.medoly.plugin.action.lyricsscraper.db.Site
import com.wa2c.android.medoly.plugin.action.lyricsscraper.db.SiteGroup
import com.wa2c.android.medoly.plugin.action.lyricsscraper.util.AppUtils
import com.wa2c.android.prefs.Prefs
import de.siegmar.fastcsv.reader.CsvReader
import kotlinx.android.synthetic.main.layout_site_item.view.*
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.GlobalScope
import kotlinx.coroutines.async
import kotlinx.coroutines.launch
import timber.log.Timber
import java.net.HttpURLConnection
import java.net.URL


/**
 * Site activity
 */
class SiteActivity : Activity() {

    private lateinit var prefs: Prefs
    private lateinit var binding: ActivityGroupBinding

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        binding = DataBindingUtil.setContentView(this, R.layout.activity_group)
        prefs = Prefs(this)

        binding.groupListView.visibility = View.VISIBLE
        binding.loadingLayout.visibility = View.INVISIBLE

        // action bar
        actionBar.setDisplayShowHomeEnabled(true)
        actionBar.setDisplayHomeAsUpEnabled(true)

        binding.groupListView.onItemClickListener = AdapterView.OnItemClickListener { _, _, position, _ ->
            val item = binding.groupListView.adapter.getItem(position)
            if (binding.groupListView.adapter is SiteGroupListAdapter) {
                openSiteList((item as SiteGroup).group_id)
            } else if (binding.groupListView.adapter is SiteListAdapter) {
                prefs.putLong(R.string.prefkey_selected_site_id, (item as Site).site_id)
                (binding.groupListView.adapter as SiteListAdapter).notifyDataSetChanged()
            }
        }
        binding.groupListView.choiceMode = AbsListView.CHOICE_MODE_SINGLE
        binding.groupListView.adapter = SiteGroupListAdapter(this)
    }


    override fun onCreateOptionsMenu(menu: Menu): Boolean {
        menuInflater.inflate(R.menu.site_list, menu)
        for (i in 0 until menu.size()) {
            menu.getItem(i).setShowAsAction(MenuItem.SHOW_AS_ACTION_IF_ROOM)
        }
        return true
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        when (item.itemId) {
            android.R.id.home -> {
                // home
                if (binding.groupListView.adapter is SiteListAdapter) {
                    openGroupList()
                } else {
                    finish()
                }
                return true
            }
            R.id.menu_update_list -> {
                GlobalScope.launch(Dispatchers.Main) {
                    val sheetId = getString(if (BuildConfig.DEBUG) R.string.sheet_id_debug else R.string.sheet_id)
                    val siteResult = async(coroutineContext + Dispatchers.Default) {
                        downloadSiteList(sheetId)
                    }
                    if (!siteResult.await()) {
                        AppUtils.showToast(applicationContext, R.string.message_renew_list_failed)
                        return@launch
                    }

                    val groupResult = async (coroutineContext + Dispatchers.Default) {
                        downloadGroupList(sheetId)
                    }
                    if (!groupResult.await()) {
                        AppUtils.showToast(applicationContext, R.string.message_renew_list_failed)
                        return@launch
                    }

                    openGroupList()
                    AppUtils.showToast(applicationContext, R.string.message_renew_list_succeeded)

                    binding.groupListView.visibility = View.VISIBLE
                    binding.loadingLayout.visibility = View.INVISIBLE
                }

                binding.groupListView.visibility = View.INVISIBLE
                binding.loadingLayout.visibility = View.VISIBLE
                return true
            }
            R.id.menu_open_sheet -> {
                val sheetId = getString(if (BuildConfig.DEBUG) R.string.sheet_id_debug else R.string.sheet_id)
                val sheetUri = Uri.parse(getString(R.string.sheet_uri, sheetId))
                val browserIntent = Intent(Intent.ACTION_VIEW, sheetUri)
                startActivity(browserIntent)
                return true
            }
        }
        return super.onOptionsItemSelected(item)
    }

    override fun dispatchKeyEvent(e: KeyEvent): Boolean {
        when (e.keyCode) {
            // back key
            KeyEvent.KEYCODE_BACK -> if (binding.groupListView.adapter is SiteListAdapter) {
                openGroupList()
                return true
            }
        }
        return super.dispatchKeyEvent(e)
    }



    /**
     * Download site list.
     */
    private fun downloadSiteList(sheetId: String): Boolean {
        var con: HttpURLConnection? = null
        try {
            val siteGid = getString(if (BuildConfig.DEBUG) R.string.sheet_site_gid_debug else R.string.sheet_site_gid)
            val siteExportUrl = getString(R.string.sheet_export_uri, sheetId, siteGid)
            val siteUrl = URL(siteExportUrl)
            con = siteUrl.openConnection() as HttpURLConnection
            con.requestMethod = "GET"
            con.doInput = true
            con.connect()
            con.inputStream.bufferedReader(Charsets.UTF_8).use {
                try {
                    val csvReader = CsvReader()
                    csvReader.setContainsHeader(true)
                    val csv = csvReader.read(it)
                    val siteList = mutableListOf<Site>()
                    for (row in csv.rows) {
                        val site = Site(
                                0,
                                row.getField("SITE_ID").toLong(),
                                row.getField("GROUP_ID").toLong(),
                                row.getField("SITE_NAME"),
                                row.getField("SITE_URI"),
                                row.getField("SEARCH_URI"),
                                row.getField("RESULT_PAGE_URI_ENCODING"),
                                row.getField("RESULT_PAGE_ENCODING"),
                                row.getField("RESULT_PAGE_PARSE_TYPE"),
                                row.getField("RESULT_PAGE_PARSE_TEXT"),
                                row.getField("LYRICS_PAGE_ENCODING"),
                                row.getField("LYRICS_PAGE_PARSE_TYPE"),
                                row.getField("LYRICS_PAGE_PARSE_TEXT"),
                                row.getField("DELAY").toLong(),
                                row.getField("TIMEOUT").toLong()
                        )
//                        val site = Site()
//                        site.site_id = row.getField("SITE_ID").toLong()
//                        site.group_id = row.getField("GROUP_ID").toLong()
//                        site.site_name = row.getField("SITE_NAME")
//                        site.site_uri = row.getField("SITE_URI")
//                        site.search_uri = row.getField("SEARCH_URI")
//                        site.result_page_uri_encoding = row.getField("RESULT_PAGE_URI_ENCODING")
//                        site.result_page_encoding = row.getField("RESULT_PAGE_ENCODING")
//                        site.result_page_parse_type = row.getField("RESULT_PAGE_PARSE_TYPE")
//                        site.result_page_parse_text = row.getField("RESULT_PAGE_PARSE_TEXT")
//                        site.lyrics_page_encoding = row.getField("LYRICS_PAGE_ENCODING")
//                        site.lyrics_page_parse_type = row.getField("LYRICS_PAGE_PARSE_TYPE")
//                        site.lyrics_page_parse_text = row.getField("LYRICS_PAGE_PARSE_TEXT")
//                        site.delay = row.getField("DELAY").toLong()
//                        site.timeout = row.getField("TIMEOUT").toLong()
                        siteList.add(site)
                    }
                    DbHelper(this@SiteActivity).renewSite(siteList)
                } catch (ex: Exception) {
                    Timber.d(ex)
                }
            }
            return true
        }  catch (e: Exception) {
            Timber.d(e)
            return false
        } finally {
            con?.disconnect()
        }
    }

    /**
     * Download group list.
     */
    private fun downloadGroupList(sheetId: String): Boolean {
        var con: HttpURLConnection? = null
        try {
            val groupGid = getString(if (BuildConfig.DEBUG) R.string.sheet_group_gid_debug else R.string.sheet_group_gid)
            val groupUri = getString(R.string.sheet_export_uri, sheetId, groupGid)
            val siteUrl = URL(groupUri)
            con = siteUrl.openConnection() as HttpURLConnection
            con.requestMethod = "GET"
            con.doInput = true
            con.connect()
            con.inputStream.bufferedReader(Charsets.UTF_8).use {
                try {
                    val csvReader = CsvReader()
                    csvReader.setContainsHeader(true)
                    val csv = csvReader.read(it)
                    val groupList = mutableListOf<SiteGroup>()
                    for (row in csv.rows) {
                        val group = SiteGroup(
                                0,
                                row.getField("GROUP_ID").toLong(),
                                row.getField("NAME"),
                                row.getField("NAME_JA")
                        )
//                        val group = SiteGroup()
//                        group.group_id = row.getField("GROUP_ID").toLong()
//                        group.name = row.getField("NAME")
//                        group.name_ja = row.getField("NAME_JA")
                        groupList.add(group)
                    }
                    DbHelper(this@SiteActivity).renewSiteGroup(groupList)
                } catch (ex: Exception) {
                    Timber.d(ex)
                }
            }
            return true
        }  catch (e: Exception) {
            Timber.d(e)
            return false
        } finally {
            con?.disconnect()
        }
    }

    /**
     * Open site list.
     */
    private fun openSiteList(groupId: Long) {
        binding.groupListView.adapter = SiteListAdapter(this, groupId)
    }

    /**
     * Open group list.
     */
    private fun openGroupList() {
        binding.groupListView.adapter = SiteGroupListAdapter(this)
    }



    companion object {
//        private const val SITE_SHEET_NAME = "SITE"
//        private const val GROUP_SHEET_NAME = "GROUP"

        /**
         * Site group adapter.
         */
        private class SiteGroupListAdapter internal constructor(context: Context) : ArrayAdapter<SiteGroup>(context, R.layout.layout_site_item) {
            init {
                addAll(DbHelper(context).selectSiteGroupList())
            }

            override fun getView(position: Int, convertView: View?, parent: ViewGroup): View {
                var itemView = convertView
                // itemView
                val holder: ListItemViewHolder
                if (itemView == null) {
                    holder = ListItemViewHolder(parent.context)
                    itemView = holder.itemView
                } else {
                    holder = itemView.tag as ListItemViewHolder
                }

                holder.bind(getItem(position))

                return itemView
            }

            /** List item view holder.  */
            private class ListItemViewHolder(context: Context) {
                val itemView = View.inflate(context, R.layout.layout_site_item, null)!!

                init {
                    itemView.tag = this
                }

                fun bind(item: SiteGroup) {
                    itemView.siteSelectRadioButton.visibility = View.GONE
                    itemView.siteLaunchImageButton.visibility = View.GONE
                    itemView.siteParamUriTextView.visibility = View.GONE
                    itemView.siteParamTitleTextView.text = item.findLocaleName(null)
                }
            }
        }

        /**
         * Site adapter.
         */
        private class SiteListAdapter internal constructor(context: Context, groupId: Long) : ArrayAdapter<Site>(context, R.layout.layout_site_item) {

            init {
                addAll(DbHelper(context).selectSiteListByGroupId(groupId))
            }

            override fun getView(position: Int, convertView: View?, parent: ViewGroup): View {
                var itemView = convertView
                val holder: ListItemViewHolder
                if (itemView == null) {
                    holder = ListItemViewHolder(parent.context)
                    itemView = holder.itemView
                } else {
                    holder = itemView.tag as ListItemViewHolder
                }

                holder.bind(getItem(position))

                return itemView
            }

            /** List item view holder.  */
            private class ListItemViewHolder(val context: Context) {
                val prefs = Prefs(context)
                val itemView = View.inflate(context, R.layout.layout_site_item, null)!!

                init {
                    itemView.tag = this
                }

                fun bind(site: Site) {
                    itemView.siteSelectRadioButton.visibility = View.VISIBLE
                    itemView.siteLaunchImageButton.visibility = View.VISIBLE
                    itemView.siteParamUriTextView.visibility = View.VISIBLE

                    itemView.siteSelectRadioButton.isChecked = (site.site_id == prefs.getLong(R.string.prefkey_selected_site_id))
                    itemView.siteSelectRadioButton.setOnTouchListener { _, event -> itemView.onTouchEvent(event) }
                    itemView.siteSelectRadioButton.isClickable = false

                    itemView.siteParamTitleTextView.text = site.site_name
                    itemView.siteParamUriTextView.text = site.site_uri
                    itemView.siteLaunchImageButton.tag = site.site_uri
                    if (!site.site_uri.isNullOrEmpty()) {
                        itemView.siteLaunchImageButton.isEnabled = true
                        itemView.siteLaunchImageButton.setOnClickListener { v ->
                            val sheetUri = Uri.parse(v.tag as String)
                            val browserIntent = Intent(Intent.ACTION_VIEW, sheetUri)
                            context.startActivity(browserIntent)
                        }
                    } else {
                        itemView.siteLaunchImageButton.isEnabled = false
                    }
                }
            }
        }
    }

}
